### PR TITLE
SOC

### PR DESCRIPTION
**Pull request by matthewha123**
Rudimentary state of charge implementation.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -59,7 +59,7 @@ ifeq ($(EVT),1)
 else
 	# include directories (absolute or relative paths to additional folders with
 	# headers, current folder is always included)
-	INC_DIRS_CROSS = inc/ inc/fsae_drivers/ ../lpc11cx4-library/lpc_chip_11cxx_lib/inc ../lpc11cx4-library/evt_lib/inc/ ../MY17/lib/MY17_Can_Library/
+	INC_DIRS_CROSS = inc/ inc/fsae_drivers/ ../Opel_EE/lpc11cx4-library/lpc_chip_11cxx_lib/inc ../Opel_EE/lpc11cx4-library/evt_lib/inc/ ../MY17/lib/MY17_Can_Library/
```
Comment (skoppula) [inline Makefile]: This is computer/directory setup specific, please remove the diff before merging.

File: inc/state_types.h [hunk 2]
```diff
@@ -39,6 +39,7 @@ typedef struct BMS_PACK_STATUS {
     uint32_t pack_current_mA;
     uint32_t pack_voltage_mV;
     int16_t max_cell_temp_dC;
+    uint32_t state_of_charge;
```
Comment (skoppula) [inline inc/state_types.h]: How are you initializing this value?

File: src/soc.c [hunk 3]
```diff
@@ -1,11 +1,46 @@
...
-    // return EEPROM_LoadCCPage_Num(0);
-    return 0;
+#define DELTA_T 50
+uint32_t SOC_Estimate(BMS_INPUT_T* bms_input, uint32_t ms_ticks) {
+
+	//rectangular integration
+	if((ms_ticks - last_tick_soc)> DELTA_T){
+		last_tick_soc = ms_ticks;
+		//units are coulombs
+		soc = (soc - (bms_input->pack_status->pack_current_mA) * DELTA_T);
```
Comment (skoppula) [inline src/soc.c]: Displayed SoC should ideally be in a 0-100 scale on the dash, so we have a human-interpretable value. Might be interested in calibrating a mapping.

**Review by skoppula — CHANGES_REQUESTED** [review]
Two things to think about and talk with me, Elliot, and Jorge with on Saturday:

1. Ideally, we'd want to be able to recalibrate the '100%' or '0%' SoC on signal -- e.g. automatically or UART command at the end of charging, or whenever the BMS hits undervoltage fault, that's 0% mapped SoC. Decide how this will be done

2. You should also think about more accurate charge estimate might be from energy readings from the current sensor (good point brought up by Elliot and Luis today). Would need to that in from CAN, and take a look at the current sensor spec (talk to Nick to see examples of past CAN logs as well to confirm the code/spec)?

**Comment by matthewha123** [comment]
I had trouble thinking of a way of scaling it 0-100 that avoided doing
division. For the first point I was thinking of the charger shutdown board
telling the bms when it is 100%.

**Pull request closed without merging**